build: Enable fuzz binary in MSVC

hebasto opened the pull request:
Closes https://github.com/bitcoin/bitcoin/issues/29760.

Suggested in https://github.com/bitcoin/bitcoin/pull/29758#issuecomment-2025593572.

hebasto commented:
Friendly ping @sipa @maflcko @sipsorcery @dergoegge ;)

hebasto commented:
> What is the issue with the bitdeque and miniscript fuzz tests?

Here is a [branch](https://github.com/hebasto/bitcoin/tree/240331-msvc-fuzz.bitdeque) with included `bitdeque.cpp` and the excerpt from its [build log](https://github.com/hebasto/bitcoin/actions/runs/8499430427/job/23280244758):
```
  bitdeque.cpp
D:\a\bitcoin\bitcoin\src\util\bitdeque.h(91,20): error C2248: 'bitdeque<128>::BITS_PER_WORD': cannot access private member declared in class 'bitdeque<128>' [D:\a\bitcoin\bitcoin\build_msvc\fuzz\fuzz.vcxproj]
  D:\a\bitcoin\bitcoin\src\util\bitdeque.h(28,26): note: see declaration of 'bitdeque<128>::BITS_PER_WORD'
  D:\a\bitcoin\bitcoin\src\test\fuzz\bitdeque.cpp(18,23): note: see declaration of 'bitdeque<128>'
  D:\a\bitcoin\bitcoin\src\util\bitdeque.h(91,20): note: the template instantiation context (the oldest one first) is
  D:\a\bitcoin\bitcoin\src\test\fuzz\bitdeque.cpp(215,17): note: see reference to class template instantiation 'bitdeque<128>::Iterator<true>' being compiled
  D:\a\bitcoin\bitcoin\src\util\bitdeque.h(89,32): note: while compiling class template member function 'bitdeque<128>::Iterator<true>::difference_type operator -(const bitdeque<128>::Iterator<true> &,const bitdeque<128>::Iterator<true> &)'
  D:\a\bitcoin\bitcoin\src\test\fuzz\bitdeque.cpp(215,17): note: see the first reference to 'operator -' in 'bitdeque_fuzz_target::<lambda_18>::operator ()'
D:\a\bitcoin\bitcoin\src\util\bitdeque.h(91,20): error C2248: 'bitdeque<128>::BITS_PER_WORD': cannot access private member declared in class 'bitdeque<128>' [D:\a\bitcoin\bitcoin\build_msvc\fuzz\fuzz.vcxproj]
  (compiling source file '../../src/test/fuzz/bitdeque.cpp')
  D:\a\bitcoin\bitcoin\src\util\bitdeque.h(28,26):
  see declaration of 'bitdeque<128>::BITS_PER_WORD'
  D:\a\bitcoin\bitcoin\src\test\fuzz\bitdeque.cpp(18,23):
  see declaration of 'bitdeque<128>'
  D:\a\bitcoin\bitcoin\src\util\bitdeque.h(91,20):
  the template instantiation context (the oldest one first) is
  	D:\a\bitcoin\bitcoin\src\test\fuzz\bitdeque.cpp(215,17):
  	see reference to class template instantiation 'bitdeque<128>::Iterator<true>' being compiled
  	D:\a\bitcoin\bitcoin\src\util\bitdeque.h(89,32):
  	while compiling class template member function 'bitdeque<128>::Iterator<true>::difference_type operator -(const bitdeque<128>::Iterator<true> &,const bitdeque<128>::Iterator<true> &)'
  		D:\a\bitcoin\bitcoin\src\test\fuzz\bitdeque.cpp(215,17):
  		see the first reference to 'operator -' in 'bitdeque_fuzz_target::<lambda_18>::operator ()'  
```

Here is a [branch](https://github.com/hebasto/bitcoin/tree/240331-msvc-fuzz.miniscript) with included `miniscript.cpp` and the excerpt from its [build log](https://github.com/hebasto/bitcoin/actions/runs/8499436657/job/23280257766):
```
  miniscript.cpp
D:\a\bitcoin\bitcoin\src\script\miniscript.h(154,13): error C2248: 'miniscript::Type::Type': cannot access private member declared in class 'miniscript::Type' [D:\a\bitcoin\bitcoin\build_msvc\fuzz\fuzz.vcxproj]
  D:\a\bitcoin\bitcoin\src\script\miniscript.h(127,5): note: see declaration of 'miniscript::Type::Type'
  D:\a\bitcoin\bitcoin\src\script\miniscript.h(122,7): note: see declaration of 'miniscript::Type'
  D:\a\bitcoin\bitcoin\src\test\fuzz\miniscript.cpp(1226,53): note: while evaluating constexpr function 'miniscript::operator ""_mst'
```

Richlilyjoy100 commented:
> Closes https://github.com/bitcoin/bitcoin/issues/29760.
> 
> Suggested in https://github.com/bitcoin/bitcoin/pull/29758#issuecomment-2025593572.
> 
> Once this PR collects concept ACKs, I'm going to open dedicated PRs to fix portability issues in the code.

Empact commented on src/test/fuzz/util/descriptor.cpp:
```diff
@@ -15,7 +15,7 @@ void MockedDescriptorConverter::Init() {
         if (IdIsCompPubKey(i) || IdIsUnCompPubKey(i) || IdIsXOnlyPubKey(i) || IdIsConstPrivKey(i)) {
             CKey privkey;
-            privkey.Set(UCharCast(key_data.begin()), UCharCast(key_data.end()), !IdIsUnCompPubKey(i));
+            privkey.Set(UCharCast(key_data.data()), UCharCast(key_data.data()) + 32, !IdIsUnCompPubKey(i));
```
How about `key_data.size()` rather than `32`?

hebasto replied:
`... + 32` is consistent with other `CKey::Set()` calls in our codebase:

https://github.com/bitcoin/bitcoin/blob/b5d21182e5a66110ce2796c2c99da39c8ebf0d72/src/bench/bip324_ecdh.cpp#L30

https://github.com/bitcoin/bitcoin/blob/b5d21182e5a66110ce2796c2c99da39c8ebf0d72/src/bench/ellswift.cpp#L20

https://github.com/bitcoin/bitcoin/blob/b5d21182e5a66110ce2796c2c99da39c8ebf0d72/src/key.cpp#L330

https://github.com/bitcoin/bitcoin/blob/b5d21182e5a66110ce2796c2c99da39c8ebf0d72/src/key.cpp#L393

https://github.com/bitcoin/bitcoin/blob/b5d21182e5a66110ce2796c2c99da39c8ebf0d72/src/key_io.cpp#L218

https://github.com/bitcoin/bitcoin/blob/b5d21182e5a66110ce2796c2c99da39c8ebf0d72/src/test/fuzz/miniscript.cpp#L57

https://github.com/bitcoin/bitcoin/blob/b5d21182e5a66110ce2796c2c99da39c8ebf0d72/src/test/miniscript_tests.cpp#L63

https://github.com/bitcoin/bitcoin/blob/b5d21182e5a66110ce2796c2c99da39c8ebf0d72/src/test/script_tests.cpp#L195

with an exception in https://github.com/bitcoin/bitcoin/blob/b5d21182e5a66110ce2796c2c99da39c8ebf0d72/src/test/orphanage_tests.cpp#L45

sipsorcery commented:
I get 14 compiler errors when building this PR with Visual Studio.

One example on line 282 of transaction_tests.cpp:

`UniValue tests = read_json(json_tests::tx_invalid);`

Gives compiler error of:

```
Error	C2664	'UniValue read_json(const std::string &)': 
cannot convert argument 1 from 'const unsigned char [53413]' to 'const std::string &'	
test_bitcoin	C:\dev\github\bitcoin\src\test\transaction_tests.cpp	282
```

It's easy enough to fix by casting the unsigned char[] to std::string but perhaps there's a deeper issue?

hebasto commented:
@sipsorcery 
> I get 14 compiler errors when building this PR with Visual Studio.

How does your setup differ from the one in the CI?

dergoegge commented:
> One example on line 282 of transaction_tests.cpp:

I believe this is unrelated to this PR, see https://github.com/bitcoin/bitcoin/issues/29051. `make clean` and rebuilding should fix it.

sipsorcery commented:
> I believe this is unrelated to this PR, see #29051. `make clean` and rebuilding should fix it.

I'm on Windows and using msvc but I did do a clean build but it didn't help.

I suspect it's down to msvc being less forgiving than gcc, probably due to a recent msvc change although I couldn't find anything in the release notes.

Apolgies for dumping a screenshot but it does show a succinct summary of the issue. This is from an attempt to compile (no linking involved) the single blockfilter_tests.cpp file.

![image](https://github.com/bitcoin/bitcoin/assets/197660/55968ab6-3438-4043-8113-6d364ea88ac2)

hebasto commented:
@sipsorcery 
> I'm on Windows and using msvc but I did do a clean build but it didn't help.

What VS / MSVC versions do you use?

hebasto commented:
> This is from an attempt to compile (no linking involved) the single blockfilter_tests.cpp file.

Does it compiles on the master branch for you?

sipsorcery commented:
> @sipsorcery
> 
> > I'm on Windows and using msvc but I did do a clean build but it didn't help.
> 
> What VS / MSVC versions do you use?

Visual Studio (including msvc) updates get automatically pushed every couple of weeks. I'm generally always pretty close to the latest stable version.

Currently Visual Studio 2022 17.8.0

```
msbuild --version
MSBuild version 17.8.3+195e7f5a3 for .NET Framework
17.8.3.51904
```

sipsorcery commented:
> > This is from an attempt to compile (no linking involved) the single blockfilter_tests.cpp file.
> 
> Does it compile on the master branch for you?

No.

I only checked that after posting the compiler error message. The msvc compiler error isn't related to this PR, it's on master as well.

hebasto commented:
> The msvc compiler error isn't related to this PR, it's on master as well.

Could you please open a separate issue then?

FWIW, I have no compile errors for the master branch @ b5d21182e5a66110ce2796c2c99da39c8ebf0d72 using VS Community 2022 17.9.5. For example, the `test_bitcoin.exe` compiles for both `Debug` and `Release` configurations:
```
> msbuild --version
MSBuild version 17.9.8+b34f75857 for .NET Framework
17.9.8.16306
> py -3 build_msvc\msvc-autogen.py
> msbuild build_msvc\bitcoin.sln -p:UseMultiToolTask=true -p:Configuration=Debug -maxCpuCount -v:minimal -noLogo -t:test_bitcoin
> msbuild build_msvc\bitcoin.sln -p:UseMultiToolTask=true -p:Configuration=Release -maxCpuCount -v:minimal -noLogo -t:test_bitcoin

hebasto commented on src/test/fuzz/util/descriptor.cpp:
```diff
@@ -15,7 +15,7 @@ void MockedDescriptorConverter::Init() {
         if (IdIsCompPubKey(i) || IdIsUnCompPubKey(i) || IdIsXOnlyPubKey(i) || IdIsConstPrivKey(i)) {
             CKey privkey;
-            privkey.Set(UCharCast(key_data.begin()), UCharCast(key_data.end()), !IdIsUnCompPubKey(i));
+            privkey.Set(UCharCast(key_data.data()), UCharCast(key_data.data()) + 32, !IdIsUnCompPubKey(i));
```
> If there is no need for it, it may be better to remove it.

Please see https://github.com/bitcoin/bitcoin/pull/29820.

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -295,3 +295,16 @@ jobs:
+        env:
+          BITCOINFUZZ: "${{ github.workspace}}\\src\\fuzz.exe"
+        shell: cmd
+        run: py -3 test\fuzz\test_runner.py --par %NUMBER_OF_PROCESSORS% --loglevel DEBUG --empty_min_time=60 %RUNNER_TEMP%\qa-assets\fuzz_seed_corpus
```
Thanks! Removed.

hebasto commented:
The branch has been reworked to address @maflcko's comments.

> ... but bool return value + exceptions doesn't make sense, does it?

Whether [`std::istream::read()`](https://en.cppreference.com/w/cpp/io/basic_istream/read) throw or not depends on how [`exceptions()`](https://en.cppreference.com/w/cpp/io/basic_ios/exceptions) is set. The suggested code does not use exceptions.

hebasto commented on src/test/fuzz/fuzz.cpp:
```diff
@@ -136,11 +135,13 @@ void initialize()
-    while ((length = read(STDIN_FILENO, buffer, 1024)) > 0) {
+    size_t length = 0;
+    do {
+        std::cin.read(reinterpret_cast<char*>(buffer), 1024);
```
Thanks! [Reworked](https://github.com/bitcoin/bitcoin/pull/29774#issuecomment-2041136905).

hebasto commented on src/test/fuzz/fuzz.cpp:
```diff
@@ -136,11 +135,13 @@ void initialize()
     uint8_t buffer[1024];
-    ssize_t length = 0;
-    while ((length = read(STDIN_FILENO, buffer, 1024)) > 0) {
+    size_t length = 0;
```
[Reworked](https://github.com/bitcoin/bitcoin/pull/29774#issuecomment-2041136905).

hebasto commented on src/test/fuzz/fuzz.cpp:
```diff
@@ -136,11 +135,13 @@ void initialize()
-    }
-    return length == 0;
+    } while (length > 0);
+    return std::cin.eof();
```
[Reworked](https://github.com/bitcoin/bitcoin/pull/29774#issuecomment-2041136905).

fanquake commented on src/test/fuzz/multiplication_overflow.cpp:
```diff
@@ -10,19 +10,33 @@
 #include <string>
 #include <vector>
 
+#if defined(__has_builtin)
```
I don't really think we want a reversion here (or in addition), and if we were reverting shouldn't the `__GNUC__` case go away?

In any case, I'd rather we just not define this for MSVC.

hebasto replied:
> I'd rather we just not define this for MSVC.

Thanks! Done.

hebasto commented:
Rebased on top of the merged bitcoin/bitcoin#29820 and bitcoin/bitcoin#29821.

The recent comments have been addressed.

hebasto commented on test/fuzz/test_runner.py:
```diff
@@ -156,7 +156,7 @@ def main():
                 '-help=1',
             ],
-            env=get_fuzz_env(target=test_list_selection[0], source_dir=config['environment']['SRCDIR']),
+            env={**os.environ, **get_fuzz_env(target=test_list_selection[0], source_dir=config['environment']['SRCDIR'])},
```
The error message is:
```
Traceback (most recent call last):
  File "D:\a\bitcoin\bitcoin\test\fuzz\test_runner.py", line 406, in <module>
    main()
  File "D:\a\bitcoin\bitcoin\test\fuzz\test_runner.py", line 110, in main
    test_list_all = parse_test_list(
                    ^^^^^^^^^^^^^^^^
  File "D:\a\bitcoin\bitcoin\test\fuzz\test_runner.py", line 392, in parse_test_list
    test_list_all = subprocess.run(
                    ^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'D:\a\bitcoin\bitcoin\src\fuzz.exe' returned non-zero exit status 3221226505.
Error: Process completed with exit code 1.
```

(see https://github.com/hebasto/bitcoin/actions/runs/8636782427/job/23677472745)

Do you think this message is meaningful enough to put in to the commit message?

hebasto replied:
On Windows, when `subprocess` sets environment variables, it effectively overrides all of them, which makes `PATH` and `COMSPEC` environment variable empty on the master branch.

hebasto replied:
Linux does not need the `COMSPEC` environment variable to spawn a new process, right?

hebasto replied:
My previous comments were a bit misleading because only [`%SystemRoot%`](https://learn.microsoft.com/en-us/windows/deployment/usmt/usmt-recognized-environment-variables) is required to be passed through into a subprocess.

From the Python [docs](https://docs.python.org/3/library/subprocess.html):
> If specified, `env` must provide any variables required for the program to execute. On Windows, in order to run a [side-by-side assembly](https://en.wikipedia.org/wiki/Side-by-Side_Assembly) the specified `env` must include a valid `SystemRoot`.

Reworked and documented.

maflcko reviewed (approved):
Haven't reviewed anything in build_msvc

lgtm ACK 37cda55b3923fa8c18059c8dfd8c61a537d6b8b5 🚗

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: lgtm ACK 37cda55b3923fa8c18059c8dfd8c61a537d6b8b5 🚗
y0gHEEPCgqu2uip6x8C4jYAbKdEI6Vq/on9xDjK2B/bXYWgzuYYONFj6PQ2s61tsMBlfEN4ZSnCJ3JTpivQlCQ==
```

</details>

hebasto committed:
fuzz: Re-implement `read_stdin` in portable way

hebasto committed:
build, msvc: Enable preprocessor conformance mode

See:
- https://learn.microsoft.com/en-us/cpp/build/reference/zc-preprocessor
- https://learn.microsoft.com/en-us/cpp/preprocessor/preprocessor-experimental-overview

Otherwise, the "traditional" MSVC preprocessor fails to parse the
`FUZZ_TARGET` and `DETAIL_FUZZ` macros because of behavior changes
highlighted in the docs mentioned above.

hebasto committed:
build, msvc: Build `fuzz.exe` binary

hebasto committed:
ci, msvc: Add "Clone fuzz corpus" step

hebasto committed:
fuzz: Pass `SystemRoot` environment variable to subprocess

See https://docs.python.org/3/library/subprocess.html

hebasto committed:
ci, msvc: Add "Run fuzz binaries" step

hebasto commented:
Addressed the recent @maflcko's comments.

hebasto commented on build_msvc/common.init.vcxproj.in:
```diff
@@ -87,7 +87,7 @@
       <WarningLevel>Level3</WarningLevel>
       <PrecompiledHeader>NotUsing</PrecompiledHeader>
-      <AdditionalOptions>/utf-8 /Zc:__cplusplus /std:c++20 %(AdditionalOptions)</AdditionalOptions>
+      <AdditionalOptions>/utf-8 /Zc:preprocessor /Zc:__cplusplus /std:c++20 %(AdditionalOptions)</AdditionalOptions>
```
An explanation has been [added](https://github.com/bitcoin/bitcoin/pull/29774#issuecomment-2063432556) to the commit message.

hebasto commented on src/test/fuzz/fuzz.cpp:
```diff
@@ -135,12 +134,13 @@ void initialize()
+        data.insert(data.end(), buffer, buffer + std::cin.gcount());
     }
-    return length == 0;
+    return true;
```
Thanks! [Implemented](https://github.com/bitcoin/bitcoin/pull/29774#issuecomment-2063432556).

sipsorcery commented:
tACK 09f5a74198c328c80539c17d951a70558e6b361e.

hebasto commented:
@sipsorcery 
> tACK [09f5a74](https://github.com/bitcoin/bitcoin/commit/09f5a74198c328c80539c17d951a70558e6b361e).

Wrong commit hash?

:)

sipsorcery commented:
> @sipsorcery
> 
> > tACK [09f5a74](https://github.com/bitcoin/bitcoin/commit/09f5a74198c328c80539c17d951a70558e6b361e).
> 
> Wrong commit hash?
> 
> :)

Every time :( I wish there were timestamps on the Github list. If only my old brain could remember to use git log next time.

tACK #18fd522ca9a74cf8690a6c9b9b729e78c6ed41fd

sipa reviewed (commented):
utACK 18fd522ca9a74cf8690a6c9b9b729e78c6ed41fd

hebasto commented:
> What is the issue with the bitdeque and miniscript fuzz tests?

The https://github.com/bitcoin/bitcoin/pull/29983 resolves the issue with the former.

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/182.

hebasto commented:
@sipa
> What is the issue with the ... miniscript fuzz tests?

The miniscript issue is resolved with your https://github.com/bitcoin/bitcoin/pull/28657 :)
